packages/prettier-plugin-java/src/printers/types-values-and-variables.js

```diff
@@ -1,81 +1,194 @@
   }
 
   numericType(ctx) {
-    return "numericType";
+    return this.visitSingle(ctx);
   }
 
   integralType(ctx) {
-    return "integralType";
+    return this.getSingle(ctx).image;
   }
 
   floatingPointType(ctx) {
-    return "floatingPointType";
+    return this.getSingle(ctx).image;
   }
 
   referenceType(ctx) {
-    return "referenceType";
+    const annotations = this.mapVisit(ctx.annotation);
+
+    const type = ctx.primitiveType
+      ? this.visit(ctx.primitiveType)
+      : this.visit(ctx.classOrInterfaceType);
+
+    const dims = this.visit(ctx.dims);
+
+    return rejectAndJoin("", [join(" ", annotations), " ", type, dims]);
   }
 
   classOrInterfaceType(ctx) {
-    return "classOrInterfaceType";
+    return this.visitSingle(ctx);
   }
 
   classType(ctx) {
-    return "classType";
+    const dots = ctx.Dots;
+
+    if (dots) {
+      const annotations = ctx.annotation;
+      const typeArguments = ctx.typeArguments;
+      const identifiers = ctx.Identifier;
+
+      const segments = [];
+      for (const dot in dots) {
+        const offset = dot.startOffset;
+
+        const currentSegmentAnnotations = [];
+        while (annotations.length > 0) {
+          const annotation = annotations[0];
+          if (annotation && annotation.startOffset < offset) {
+            currentSegmentAnnotations.push(this.visit(annotation));
+            annotations.pop();
+          } else {
+            break;
+          }
+        }
+        const currentSegmentIdentifier = identifiers.pop().image;
+
+        let currentSegmentTypeArgument = undefined;
+        const typeArgument = typeArguments[0];
+        if (typeArgument && typeArgument.startOffset < offset) {
+          currentSegmentTypeArgument = this.visit(typeArgument);
+          typeArguments.pop();
+        }
+
+        segments.push(
+          concat(
+            join(" ", currentSegmentAnnotations),
+            currentSegmentIdentifier,
+            currentSegmentTypeArgument
+          )
+        );
+      }
+
+      return join(".", segments);
+    }
+
+    const annotations = this.mapVisit(ctx.annotation);
+    const identifier = ctx.Identifier[0].image;
+    const typeArguments = this.mapVisit(ctx.typeArguments);
+
+    return rejectAndJoin("", [
+      join(" ", annotations),
+      identifier,
+      typeArguments
+    ]);
   }
 
   interfaceType(ctx) {
-    return "interfaceType";
+    return this.visitSingle(ctx);
   }
 
   typeVariable(ctx) {
-    return "typeVariable";
+    const annotations = this.mapVisit(ctx.annotation);
+    const identifier = this.getSingle(ctx).image;
+
+    return join(" ", [join(" ", annotations), identifier]);
   }
 
   dims(ctx) {
+    // TODO
     return "dims";
   }
 
   typeParameter(ctx) {
-    return "typeParameter";
+    const typeParameterModifiers = this.mapVisit(ctx.typeParameterModifier);
+    const typeParameterModifiersOutput =
+      typeParameterModifiers.length > 0
```

Comment:
either its undefined or with length > 0
if its undefined there would be a runtime error that .length property does not exist on undefined.

optional productions are undefined in the CST